### DISCUSSION TITLE
Use temporary token for backend DB

### DISCUSSION THREAD

**Discussion by albertocalderari** [apache/airflow#30283] [General]
Hi, I am using pod identity and temporary tokens to access the backend DB.
So far I went int a rabbit hole of issues as the connection seems to be defined slighly differently in different places.

My first attempt was to "quick and dirty" replace the orm handler with:

```python
def amend_connection(cparams):
    if conf.getboolean("aws", "use_pod_identity"):
        log.info(f'connecting user {cparams["user"]} to {cparams["host"]}:{cparams["host"]} using pod identity')
        client = boto3.client(
            "rds",
            region_name=conf.get_mandatory_value("aws", "region"),
        )
        token = client.generate_db_auth_token(
            DBHostname=cparams["host"],
            Port=cparams["port"],
            DBUsername=cparams["user"],
        )
        cparams["password"] = token
    else:
        log.info(f'connecting  {cparams["user"]} using user/password')


def setup_event_handlers(engine):
    """Setups event handlers."""
    from airflow.models import import_all_models
    event.listen(sqlalchemy.orm.mapper, "before_configured", import_all_models, once=True)

    @event.listens_for(engine, "connect")
    def connect(dbapi_connection, connection_record):
        connection_record.info["pid"] = os.getpid()

    if engine.dialect.name == "sqlite":
        @event.listens_for(engine, "connect")
        def set_sqlite_pragma(dbapi_connection, connection_record):
            cursor = dbapi_connection.cursor()
            cursor.execute("PRAGMA foreign_keys=ON")
            cursor.close()

    # this ensures coherence in mysql when storing datetimes (not required for postgres)
    if engine.dialect.name == "mysql":
        @event.listens_for(engine, "connect")
        def set_mysql_timezone(dbapi_connection, connection_record):
            cursor = dbapi_connection.cursor()
            cursor.execute("SET time_zone = '+00:00'")
            cursor.close()

        @event.listens_for(engine, "do_connect")
        def provide_token(dialect, conn_rec, cargs, cparams):
            amend_connection(cparams)

    if engine.dialect.name == "postgresql":
        @event.listens_for(engine, "do_connect")
        def provide_token(dialect, conn_rec, cargs, cparams):
            amend_connection(cparams)
```
           
This works on the scheduler and workers but fail the db_migrations as the DB migrations soed not use the `setup_event_handlers ` but uses the connection directly. givinge me a few headaches.
Now for this to pass I have set these in the secret-config-envs.yaml section of the helm chart:

```helm
  {{- if .Values.externalDatabase.awsRds.enabled }}
  DATABASE_REGION: {{ .Values.externalDatabase.awsRds.region | toString | b64enc | quote }}
  DATABASE_PASSWORD_CMD: {{ `python3 -c "import boto3; import urllib; client = boto3.client('rds',region_name='${DATABASE_REGION}'); token=client.generate_db_auth_token(DBHostname='${DATABASE_HOST}',Port='${DATABASE_PORT}',DBUsername='${DATABASE_USER}'); print(urllib.parse.quote_plus(token));"` | b64enc | quote }}
  ## needed as the temporary token expres
  AIRFLOW__DATABASE__SQL_ALCHEMY_POOL_RECYCLE: {{ .Values.externalDatabase.awsRds.connectionExpirySeconds | toString | b64enc | quote }}
  AIRFLOW__AWS__REGION: {{ .Values.externalDatabase.awsRds.connectionExpirySeconds | toString | b64enc | quote }}
  {{- else }}
  DATABASE_PASSWORD_CMD: {{ `echo "${DATABASE_PASSWORD}" | python3 -c "import urllib.parse; encoded_pass = urllib.parse.quote(input()); print(encoded_pass)"` | b64enc | quote }}
  {{- end }}
  ```
  
  This works well for the scheduler and workers and pass db_migration.
  Now, the webserver unfortunately uses the sqlalchemy conn from config which is stale as DATABASE_PASSWORD_CMD gets called only once, and after 1h the webserver dies and never heals. I sincerely can't believe that I'm the first person doing this, and likely I am doing something very wrong.
From the engineering standpoint I think providing access to the  `setup_event_handlers` and making sure all sqlalchem connections go through it would be ideal. So that any credential provider can be injected.

**Comment by potiuk** [comment]
I believe on webserver you have gunicorn workers that get periodically restarted and yoy can define their time of live - look at the configuration (`worker_refresh_interval'. Every time they get restarated they will refresh their connecitons. You do not need to do anything special there. It is set to 6000 s by default which is more than 3 hours so you should likely shorten it.

**Comment by albertocalderari** [comment]
> I believe on webserver you have gunicorn workers that get periodically restarted and yoy can define their time of live - look at the configuration (`worker_refresh_interval'. Every time they get restarated they will refresh their connecitons. You do not need to do anything special there. It is set to 6000 s by default which is more than 3 hours so you should likely shorten it.
> 
> 
> 
> 

Let me try this and get back to you. How would this affect long running workers BTW?

**Reply by potiuk** [reply]
It is unrelated. 

But apparently you have solution to workers already - so just continue doing so. This is a very specific requirements and seems that you found a good solution for your deployment case. 

But yes that might be a good idea to have a better support for it -  if you would like to improve Airflow to support it better - you are always welcome to propose it in devlist and submit PR(s) implementing it. That's the usual process that is followed in open-source Airlfow (it is created by almost 2400 contrinutors - most of them people like you who see a missing feature and would like to add it). You seem to be knowledgeable enough to lead such proposal and implementaiton I think.

Here is the contributing guideline  https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst - happy to guide you in the process if you have questions.

**Reply by albertocalderari** [reply]
Yeah sorry, was about to reply as I realized my question was poorly formulated LOL.
So yeah the aidea behind this discussion is, should I PR a change where the db gets in stantiated with sume custom `setup_event_handlers`.
I don't like the idea of a solution for initdb and one for the rest :D
I'll propse and implement, should I use `airflow_local_settings.py` to store `setup_event_handlers`? I'm  not too familiar with it and struggled to find much info.

**Reply by potiuk** [reply]
This is quite substantial change on how airflow manages connections. Exposing it to external world as something more than "my local deployment configuration I'v dene" is big enough change that requires devlist discussion https://lists.apache.org/list.html?dev@airflow.apache.org and possibly even AIP (Airflow Improvement Proposal - https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals) if people from the community would like to better understand the impact and what it means from Airflow future point of view (i.e. will that block us from making some changes in the future or make it more difficult, or maybe it clashes with some other initiatives that are running).

So in order to make such a change you need to start a discussion - with proposal described, convince the community that it is a good thing to do (and why) and hopefully implement it and lead to completion. That's the way it works here.

**Reply by albertocalderari** [reply]
Fair enough, I will articulate my thoughts there. Thanks

**Reply by albertocalderari** [reply]
I'll feature token Identity and propose the "external connection" in another feature as I believe token Identity is a must but externally managed connections could be a bit more of a hurdle to get through

**Reply by albertocalderari** [reply]
@potiuk I assume all i have to do is send an email to dev@airflow.apache.org

**Reply by potiuk** [reply]
yes

**Reply by albertocalderari** [reply]
kk, done it, I can't see it in the list but I assume it's my issues.

**Comment by albertocalderari** [comment]
@potiuk any update on thin BTW?

**Reply by potiuk** [reply]
I think you should follow up on the list, not here. Maybe no-one is interested or maybe simply Easter in-between did not help with people looking at it. Generally speaking you should expect that people respond on devlist (and here) when they have time to look at it and when they pick interest (they are all volunteers). 

And it's in your best interest to follow up there when you started a thread (but you have to be considerate of people's time and give them sufficient time to respond). I will not be proxying your question and proposal - it's yours to bring and convince people it's worth looking at. 

Just subscribe to the devlist, follow up,  be persistent but considerate of people's time at the same time - knowing that they are all volunteers and responding when they can and feel like they can bring something to a discussion. Present appealing csae that will pick some interest of people (and eventually likely you will have to raise and lead to completion a PR that implements what you think will be the result of discussion that follows from there. Sometimes it takes days or weeks for people to respond there so you also have to be patient. That's how things work here.

You can also prepare a PR showing up your proposal - this always helps if you play "show me the code" - but you have to be prepared that your idea will be rejected, this happens often (we see code as liability rather than asset so throwing out a code that is written is perfectly normal when consensus cannot be achieved).

And BTW. if you are unsure if your message was delivered https://lists.apache.org/list.html?dev@airflow.apache.org is the place where you can check the whole devlist archive.

**Reply by albertocalderari** [reply]
Yeah, I can't see it there, hence why I believe it might not have gone through. I'll try again later today. Thanks
